### ISSUE TITLE
dependencies correct?

### ISSUE THREAD

**Issue by bertsky** [OCR-D/ocrd_calamari#6]
Is it possible that `calamari-ocr` and `tensorflow-gpu` is too broad a range of versions supported by this repo?

With Tensorflow 2.0 I get on Calamari 0.3.1:
```
File "ocrd_calamari/recognize.py", line 31, in _init_calamari
    self.predictor = MultiPredictor(checkpoints=checkpoints)
  File "calamari_ocr/ocr/predictor.py", line 203, in __init__
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 203, in <listcomp>
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 100, in __init__
    ckpt = Checkpoint(checkpoint, auto_update=self.auto_update_checkpoints)
  File "calamari_ocr/ocr/checkpoint.py", line 35, in __init__
    self.checkpoint = json_format.Parse(f.read(), CheckpointParams())
  File "protobuf/json_format.py", line 406, in Parse
    return ParseDict(js, message, ignore_unknown_fields)
  File "protobuf/json_format.py", line 421, in ParseDict
    parser.ConvertMessage(js_dict, message)
  File "protobuf/json_format.py", line 452, in ConvertMessage
    self._ConvertFieldValuePair(value, message)
  File "protobuf/json_format.py", line 552, in _ConvertFieldValuePair
    raise ParseError('Failed to parse {0} field: {1}'.format(name, e))
google.protobuf.json_format.ParseError: Failed to parse model field: Failed to parse network field: Failed to parse backend field: Message type "BackendParams" has no field named "shuffleBufferSize".
 Available Fields(except extensions): <MessageFields sequence>
```

... and on Calamari 0.3.5:
```
File "ocrd_calamari/recognize.py", line 31, in _init_calamari
    self.predictor = MultiPredictor(checkpoints=checkpoints)
  File "calamari_ocr/ocr/predictor.py", line 220, in __init__
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 220, in <listcomp>
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 106, in __init__
    backend = create_backend_from_proto(self.network_params, restore=self.checkpoint, processes=processes)
  File "calamari_ocr/ocr/backends/factory.py", line 28, in create_backend_from_proto
    from calamari_ocr.ocr.backends.tensorflow_backend.tensorflow_backend import TensorflowBackend
  File "calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_backend.py", line 4, in <module>
    from calamari_ocr.ocr.backends.tensorflow_backend.tensorflow_model import TensorflowModel
  File "calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_model.py", line 3, in <module>
    import tensorflow.contrib.cudnn_rnn as cudnn_rnn
  File "tensorflow/contrib/__init__.py", line 33, in <module>
    from tensorflow.contrib import cudnn_rnn
  File "tensorflow/contrib/cudnn_rnn/__init__.py", line 34, in <module>
    from tensorflow.contrib.cudnn_rnn.python.layers import *
  File "tensorflow/contrib/cudnn_rnn/python/layers/__init__.py", line 23, in <module>
    from tensorflow.contrib.cudnn_rnn.python.layers.cudnn_rnn import *
  File "tensorflow/contrib/cudnn_rnn/python/layers/cudnn_rnn.py", line 20, in <module>
    from tensorflow.contrib.cudnn_rnn.python.ops import cudnn_rnn_ops
  File "tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py", line 20, in <module>
    from tensorflow.contrib.eager.python import checkpointable_utils
  File "tensorflow/contrib/eager/python/checkpointable_utils.py", line 38, in <module>
    from tensorflow.python.training import checkpointable as core_checkpointable
ImportError: cannot import name 'checkpointable'
```

...when trying to recognize with the checkpoint files provided by @mikegerber

**Comment by mikegerber** [comment]
I'll investigate. For now try installing tensorflow-gpu 1.14 before installing orcd_calamari, that should work and will fall back to CPU. (In fact I am using it on CPU because of a TensorFlow bug affecting my GPU)

**Comment by mikegerber** [comment]
@bertsky Can you also provide the full command line you use? I see the [potential for confusion](https://github.com/Calamari-OCR/calamari/issues/130) :)

**Comment by bertsky** [comment]
> @bertsky Can you also provide the full command line you use? I see the [potential for confusion](https://github.com/Calamari-OCR/calamari/issues/130) :)

Thanks for pointing me there. Looking at the [code of calamari-predict](https://github.com/Calamari-OCR/calamari/blob/a16ce94e24e618da79fbab0fe9695d311482a046/calamari_ocr/scripts/predict.py#L33), I think the way `ocrd_calamari` defines/uses the `checkpoint` parameter is wrong or at least misleading, because is not the same as the equally named parameter of `calamari-predict`. It is the result of globbing all JSON files into a list and then removing the `.json` extension which is required here.

Moreover, I think it would be useful to not rely on the CWD for relative paths, because this is not reliable across the many layers (e.g. a script which calls `ocrd-calamari-recognize` which calls `calamari.ocr.MultiPredictor`). Instead, like `TESSDATA` for Tesseract one could define an installation prefix via setuptools (overridable via environment variable), or simply use `os.path.dirname(os.path.abspath(__file__))` as reference, i.e. the directory where ocrd_calamari is installed. Absolute pathnames should stay untouched, however.

**Comment by mikegerber** [comment]
>  Looking at the [code of calamari-predict](https://github.com/Calamari-OCR/calamari/blob/a16ce94e24e618da79fbab0fe9695d311482a046/calamari_ocr/scripts/predict.py#L33), I think the way `ocrd_calamari` defines/uses the `checkpoint` parameter is wrong or at least misleading, because is not the same as the equally named parameter of `calamari-predict`. It is the result of globbing all JSON files into a list and then removing the `.json` extension which is required here.

I don't see your point? I use ocrd_calamari exactly in the way I use calamari-predict, by pointing it to the '*.ckpt.json' files. With calamari-predict I use the shell's globbing mechanism to point it to `*.ckpt.json` instead of listing the `{0,1,2,3,4}.ckpt` prefixes by hand; With ocrd_calamari I give it the glob pattern right away.

**Comment by mikegerber** [comment]
> Moreover, I think it would be useful to not rely on the CWD for relative paths, because this is not reliable across the many layers (e.g. a script which calls `ocrd-calamari-recognize` which calls `calamari.ocr.MultiPredictor`). Instead, like `TESSDATA` for Tesseract one could define an installation prefix via setuptools (overridable via environment variable), or simply use `os.path.dirname(os.path.abspath(__file__))` as reference, i.e. the directory where ocrd_calamari is installed. Absolute pathnames should stay untouched, however.

I must say I currently do not understand where you see reliance on the CWD wrt the checkpoint files, but it does not seem to be related to this issue and opened #7.

**Comment by bertsky** [comment]
> I use ocrd_calamari exactly in the way I use calamari-predict, by pointing it to the '*.ckpt.json' files. With calamari-predict I use the shell's globbing mechanism to point it to `*.ckpt.json` instead of listing the `{0,1,2,3,4}.ckpt` prefixes by hand; With ocrd_calamari I give it the glob pattern right away.

Ok, so I overlooked that `ocrd_calamari.recognize` itself already does a `glob()` and that inside `calamari.ocr.MultiPredictor`, [Checkpoint appends .json again](https://github.com/Calamari-OCR/calamari/blob/a16ce94e24e618da79fbab0fe9695d311482a046/calamari_ocr/ocr/checkpoint.py#L14). So yes, the result should be the same (and you can pass a glob pattern to both CLIs). Sorry!

**Comment by bertsky** [comment]
> I must say I currently do not understand where you see reliance on the CWD wrt the checkpoint files, but it does not seem to be related to this issue and opened #7.

Thanks for splitting off – I will try to answer there.

**Comment by bertsky** [comment]
> For now try installing tensorflow-gpu 1.14 before installing orcd_calamari

That changes the error to:
```
File "ocrd_calamari/recognize.py", line 31, in _init_calamari
    self.predictor = MultiPredictor(checkpoints=checkpoints)
  File "calamari_ocr/ocr/predictor.py", line 220, in __init__
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 220, in <listcomp>
    data_preproc=data_preproc, processes=processes) for cp in checkpoints]
  File "calamari_ocr/ocr/predictor.py", line 106, in __init__
    backend = create_backend_from_proto(self.network_params, restore=self.checkpoint, processes=processes)
  File "calamari_ocr/ocr/backends/factory.py", line 28, in create_backend_from_proto
    from calamari_ocr.ocr.backends.tensorflow_backend.tensorflow_backend import TensorflowBackend
  File "calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_backend.py", line 1, in <module>
    import tensorflow as tf
  File "tensorflow/__init__.py", line 28, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "tensorflow/python/__init__.py", line 83, in <module>
    from tensorflow.python import keras
  File "tensorflow/python/keras/__init__.py", line 26, in <module>
    from tensorflow.python.keras import activations
  File "tensorflow/python/keras/activations/__init__.py", line 22, in <module>
    from tensorflow.python.keras._impl.keras.activations import elu
  File "tensorflow/python/keras/_impl/keras/__init__.py", line 21, in <module>
    from tensorflow.python.keras._impl.keras import activations
  File "tensorflow/python/keras/_impl/keras/activations.py", line 23, in <module>
    from tensorflow.python.keras._impl.keras import backend as K
  File "tensorflow/python/keras/_impl/keras/backend.py", line 38, in <module>
    from tensorflow.python.layers import base as tf_base_layers
  File "tensorflow/python/layers/base.py", line 25, in <module>
    from tensorflow.python.keras import backend
  File "tensorflow/python/keras/backend/__init__.py", line 22, in <module>
    from tensorflow.python.keras._impl.keras.backend import abs
ImportError: cannot import name 'abs'
```

I have:
- `calamari-ocr==0.3.5`
- `tensorflow-gpu==1.14.0`
- `keras==2.2.4`

This was the command I have been using each time:
```bash
ocrd-calamari-recognize -I OCR-D-GT-SEG-LINE-BINPAGE-sauvola-DENOISE-ocropy-DESKEW-ocropy-CLIP-DESKEW-tesseract-RESEG-DEWARP -O OCR-D-OCR-CALA-gt4histocr-BINPAGE-sauvola-DENOISE-ocropy-DESKEW-ocropy-CLIP-DESKEW-tesseract-RESEG-DEWARP -p OCR-D-OCR-CALA-gt4histocr-BINPAGE-sauvola-DENOISE-ocropy-DESKEW-ocropy-CLIP-DESKEW-tesseract-RESEG-DEWARP.json
```

And this is the content of the parameter JSON:
```json
{ "checkpoint" : "env3/share/calamari/GT4HistOCR/*.ckpt.json" }
```

**Comment by bertsky** [comment]
(It does not get better with `keras==2.3.1`.) Anything else I could try out?

**Comment by mikegerber** [comment]
I do not even have full keras installed, only (via dependency):

~~~
Keras-Applications     1.0.8     
Keras-Preprocessing    1.1.0  
~~~

I've tested these combinations:
* calamari-ocr==0.3.5 + tensorflow-gpu==1.14.0
* calamari-ocr==0.3.5 + tensorflow-gpu==1.15.0

My hunch is there is something wrong with your TensorFlow installation, did you create a fresh virtualenv for this? Other people are reporting [similar symptoms](https://github.com/tensorflow/probability/issues/46).

**Comment by bertsky** [comment]
I see. Thanks for pointing me to that Tensorflow issue. I had another `tensorflow` from a custom build in the venv. So I removed that and reinstalled. Now it works!

So this was _not_ a Calamari issue. But your updated requirements also helps – thanks!

BTW, I just tried with `calamari-ocr==1.0` from upstream (not published yet), which relies on Tensorflow 2.0. I don't get the above errors, just an error about the old models:

```
Due to a update to tensorflow 2.0, the weights can not be converted yet. A retraining is required.
```

**Comment by mikegerber** [comment]
> BTW, I just tried with `calamari-ocr==1.0` from upstream (not published yet), which relies on Tensorflow 2.0. I don't get the above errors, just an error about the old models:
> 
> ```
> Due to a update to tensorflow 2.0, the weights can not be converted yet. A retraining is required.
> ```

I'll open another issue for this.
